### PR TITLE
482/define validity date

### LINKED ISSUE #482: Allow to define a validity date
By default we set validity to 30 minutes
Add input for arbitrary validity

**Estimation: 2.5**

### PR DESCRIPTION
**Pull request by W3stside**
Closes #482 

Allows user to define validity date
`expires=0 or expires=` triggers unlimited order
negative values excluded, antthing below 5 and above 0 defaulted back to 30

Has fixes from 482.5 that fixed the form bug and checkbox bug

### CODE REVIEW

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr509--dexreact.review.gnosisdev.com

**Comment by alfetopito** [comment]
Build failure due to `TS6133: 'MAX_BATCH_ID' is declared but its value is never read.`

File: src/hooks/usePlaceOrder.ts [hunk 1]
```diff
@@ -81,7 +81,8 @@ export const usePlaceOrder = (): Result => {
             userAddress,
             buyTokenId,
             sellTokenId,
-            validUntil: validUntil || batchId + DEFAULT_ORDER_DURATION,
+            // validUntil time or max time
+            validUntil: validUntil ? batchId + validUntil : MAX_BATCH_ID, //batchId + DEFAULT_ORDER_DURATION,
```
Comment (alfetopito) [inline src/hooks/usePlaceOrder.ts]: Can you update/remove the comment at the end of the line?

File: src/components/TradeWidget/OrderValidity.tsx [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  return (
+    <Wrapper>
+      <h3>Expiration time:</h3>
+      <InputBox>
+        <div className="main-input-container">
+          <input
+            className={className}
+            name={inputId}
+            type="number"
+            step="30"
```
Comment (alfetopito) [inline src/components/TradeWidget/OrderValidity.tsx]: I think the step should be 5min, given the batch size. Although it wouldn't be precise.

File: src/hooks/usePlaceOrder.ts (hunk 1 above)
Comment (W3stside) [reply src/hooks/usePlaceOrder.ts]: yep

File: src/components/TradeWidget/OrderValidity.tsx (hunk 2 above)
Comment (W3stside) [reply src/components/TradeWidget/OrderValidity.tsx]: kk good idea

File: src/components/TradeWidget/OrderValidity.tsx [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+  &.error {
+    color: red;
+  }
+  &.warning {
+    color: orange;
+  }
+`
+
+const validInputPattern = new RegExp(/^\d+\.?\d*$/) // allows leading and trailing zeros
```
Comment (Velenir) [inline src/components/TradeWidget/OrderValidity.tsx]: This would allow for `234324.`
Is that alright?

File: src/hooks/useQuery.ts [hunk 4]
```diff
@@ -6,21 +6,33 @@ import { useMemo } from 'react'
...
+
+export function useQuery(): { sellAmount: string; buyAmount: string; validUntil?: string } {
   const query = new URLSearchParams(useLocation().search)
 
   // eslint-disable-next-line react-hooks/exhaustive-deps
   const sellAmount = useMemo(() => sanitizeInput(query.get('sell')), [])
   // eslint-disable-next-line react-hooks/exhaustive-deps
   const buyAmount = useMemo(() => sanitizeInput(query.get('buy')), [])
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const validUntil = useMemo(() => sanitizeNegativeInput(query.get('expires'), '30'), [])
```
Comment (Velenir) [inline src/hooks/useQuery.ts]: I'm a bit concerned with empty deps
What if user goes back and forth beteen 
```
https://dex.staging.gnosisdev.com/trade/WETH-USDC?sell=0&buy=0
and
https://dex.staging.gnosisdev.com/trade/USDC-WETH?sell=0&buy=0
```
by just clicking :arrow_up_down: arrows

Comment (W3stside) [reply src/hooks/useQuery.ts]: I don't understand this one, what is the issue with the empty deps and switching?

Comment (Velenir) [reply src/hooks/useQuery.ts]: `useQuery()` would return the same even if `?sell|buy` changes

Comment (Velenir) [reply src/hooks/useQuery.ts]: I will try to reproduce it first

Comment (Velenir) [reply src/hooks/useQuery.ts]: #515

File: src/components/TradeWidget/OrderValidity.tsx [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            name={inputId}
+            type="number"
+            step="5"
+            disabled={isDisabled || unlimited}
+            required
+            ref={register({
+              pattern: { value: validInputPattern, message: 'Expiration time cannot be negative' },
+              validate: { positive: validatePositive },
+            })}
+            onChange={handleChange}
```
Comment (Velenir) [inline src/components/TradeWidget/OrderValidity.tsx]: Add `onFocus={(e): void => e.target.select()}` to bring in line with #479

File: src/components/TradeWidget/OrderDetails.tsx [hunk 6]
```diff
@@ -64,7 +67,16 @@ const OrderDetails: React.FC<Props> = ({
 
       <dd>Expiration date:</dd>
       <dt>
-        <Highlight>30 min</Highlight>
+        <Highlight>
+          {+validUntil == 0
+            ? 'Unlimited'
+            : `~
+          ${(+validUntil / 60)
```
Comment (anxolin) [inline src/components/TradeWidget/OrderDetails.tsx]: Should we move this to a util in formatUtil?

File: src/hooks/useQuery.ts [hunk 7]
```diff
@@ -6,22 +6,31 @@ import { useMemo } from 'react'
...
 
-export function useQuery(): { sellAmount: string; buyAmount: string } {
-  const query = new URLSearchParams(useLocation().search)
+/**
+ * Prevents invalid NEGATIVE numbers from being inserted by hand in the URL
+ *
+ * @param value Input from URL
+ */
+function sanitizeNegativeInput(value?: string | null, defaultValue = '0'): string {
+  return (Number(value) === 0 || Number(value) >= 5) && value ? value : defaultValue
```
Comment (anxolin) [inline src/hooks/useQuery.ts]: I don't understand this expression:
- Why 5?
- It looks `Number(value)` can be factorized
- Why `&& value` is checked after you actually use the `value`. Maybe you want to invert the order of the AND claused

File: src/hooks/usePlaceOrder.ts [hunk 8]
```diff
@@ -74,30 +74,42 @@ export const usePlaceOrder = (): Result => {
           exchangeApi.getCurrentBatchId(networkId),
         ])
 
-        if (sellTokenId === buyTokenId) {
-          toast.error('Cannot place order. Tokens are the same')
+        if (sellTokenId !== 0 || buyTokenId !== 0) {
+          log('sellTokenId, buyTokenId, batchId', sellTokenId, buyTokenId, batchId, sellToken.address, buyToken.address)
```
Comment (alfetopito) [inline src/hooks/usePlaceOrder.ts]: Careful there with the merge. This is old code before the unregistered token fix!

Comment (W3stside) [reply src/hooks/usePlaceOrder.ts]: ah wtf? i thought sth weird happened, I'll check

**Pull request merged**